### ISSUE TITLE
PAL_GetResourceString not working under Linux

### ISSUE THREAD

**Issue by milizhang** [dotnet/coreclr#841]
### Issue

Currently under Linux, when unixcorerun get an uncatched exception, it would print out a message like below:

```
$./corerun HelloWorld.exe
17FC System.IO.FileNotFoundException: 1A88
at System.Console.WriteLine(String value)
at Program.DrawWindows()
at Program.Main(String[] args)
```

Instead of the human-readable messages, it prints out some hex values instead.
### Analysis

After some tracing I found that the error message is translated by PAL_GetResourceString, which eventually calls dgettext(3) to get the proper message.
However, setlocale(3) is not called anywhere during execution. The program is using the default "C" locale, and because of that dgettext(3) would not refer the .mo files as it does not know which language/locale to use.

After adding `setlocale(LC_ALL, "")` in unixcorerun, I somehow fixed this problem, at least under en_US.UTF-8 locale.

```
$ ./corerun HelloWorld.exe

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.IO, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. Error 2
   at System.Console.WriteLine(String value)
   at Program.DrawWindows()
   at Program.Main(String[] args)
```
### Fixes

Now here's some questions before a formal fix:
- Where should setlocale(3) be called?
  - In corerun? Or somewhere inside CoreCLR?
  - When user changes the culture in managed program, should we call setlocale again?
- As the "original message (the hex values which serves as the search key)" is not human readable, what should we do if the user uses an unsupported locale?
  - Manually fall back to English?
  - Or leave the unreadable message as it-is?
  - First of all, is it a good practice to even use the hex ID as the "original message"?
    - As far as I know many program that utilizes gettext/libintl uses the English for the "original message" (which means they pass English messages to gettext function), so that even if a translation is not available there will still be readable messages.

**Comment by janvorli** [comment]
It is strange, it looks like someone has removed the call to set the LC_MESSAGES part of the locale to en_US. I wonder how come it even works on Linux without it (if it still works at all)
In my original commit of the gettext stuff, there was this line in the src/pal/src/locale/unicode.cpp in the CODEPAGEInit function:
```c
  // Set the locale for string resources to en_US  
  // UNIXTODO: After we add localized resources, change this to check the current   
  // locale and set it to en_US only if we don't have resources for the current locale.  
  setlocale(LC_MESSAGES, "en_US.UTF-8");  
```

**Comment by janvorli** [comment]
Ah, so they don't work on Linux now either. I'll fix it and then it should work on FreeBSD too.
@milizhang Regarding your questions, we should fall back to English if we don't have messages for the given locale. 
It is true that the "standard" gettext usage is using the English string as the "key". I did it differently so that we don't have to have the english messages compiled into the coreclr. The whole coreclr refers to those strings using the resource ids that we are using as the keys, so we would have to add lookup of a message from the id to the code and then use the retrieved string as a key to call the gettext.
I need to experiment with that, but it seems to me that if we set the locale for messages before calling the gettext API and restore it to what it was before after that, we should be safe and not to interfer even in apps that would need to use the original locale.

**Comment by milizhang** [comment]
It is because that the whole CODEPAGEInit was removed from the file.

Also if we use the hex value to support localized resources, we need to manually check the presence of the MO file, as we need to handle the fallback case manually.
To me that does not seems to be a "common" usage of gettext, though.

**Comment by janvorli** [comment]
I didn't say it is a common usage. But adding those english strings as compiled in strings to the coreclr would be ugly and it would require double lookup (one in our code and the other in the gettext). 
I don't think we need to do manual check of the mo file presence. It would be enough to call the gettext and if we get back the same hex id, we need to fall back.

**Comment by milizhang** [comment]
Oh, that is also a good idea. Maybe we can do that in PAL_GetResourceString, so that it would also support partially translated locales.

**Comment by janvorli** [comment]
Yes, that's what I'd like to do.

**Comment by ezephir** [comment]
This is implemented in 63e3fda505d2eebf89758339d5c970469f822697.
